### DISCUSSION TITLE
How to create not null not required field for jsonschema?

### DISCUSSION THREAD

**Discussion by JMarkin** [pydantic/pydantic#8845] [Question]
Hello, for example I don't have default value for this field and want mark in jsonschema it is not nullable.

Currently I use this hack
```python
def und():
  return PydanticUndefined
class A(BaseModel):
  not_required_not_nullable: str = Field(default_factory=und)
```
But it creates many warnings.
Is there a better solution?

**Comment by rafalkrupinski** [comment]
How exactly would you like the field to behave? Specifically what should happen when the field is not set?
In typed python (mypy) the general assumption is that the field is always there, either set to a value or None.

In case of `TypedDict`s there's `Required` and `NotRequired`, but deliberately not supported for pydantic models - #7712

A field is not required when it has a default value, but it adds the default value to the schema.
```python
class A(BaseModel):
    a: str=None
```
```yaml
properties:
  a:
    default: null
    title: A
    type: string
title: A
type: object
```
(This generates a valid JSON schema, the default value doesn't have to match the schema)

You can also customize the schema generation with `pydantic.GenerateJsonSchema`.

**Reply by JMarkin** [reply]
@rafalkrupinski thank you, I want generate json schema without `default: null`, because tools like schematesis   require valid schema.
My final result it generate json schema from parent model but all fields optional and not null(if possible).

Can you give example how change `GenerateJsonSchema` and use it for specific model? I tried but don't understand how.
For me mypy checking doesn't require.

**Reply by rafalkrupinski** [reply]
>> (This generates a valid JSON Schema, the default value doesn't have to match the schema)

> I want generate json schema without default: null, because tools like schematesis require valid schema.

🤷

> Can you give example how change GenerateJsonSchema and use it for specific model?

I've never used it.
